src/highdicom/seg/sop.py

```diff
@@ -944,3 +966,1458 @@ def _encode_pixels(self, planes: np.ndarray) -> bytes:
+            raise ValueError(
+                'Segment numbers array does not match the shape of the '
+                'seg frames matrix.'
+            )
+
+        if (
+            segment_numbers.min() < 1 or
+            segment_numbers.max() > self.number_of_segments
+        ):
+            raise ValueError(
+                'Segment numbers array contains invalid values.'
+                '-1.'
+            )
+
+        # Initialize empty pixel array
+        pixel_array = np.zeros(
+            (
+                seg_frames_matrix.shape[0],
+                self.Rows,
+                self.Columns,
+                seg_frames_matrix.shape[1]
+            ),
+            self.pixel_array.dtype
+        )
+
+        # Loop through output frames
+        for out_frm, frm_row in enumerate(seg_frames_matrix):
+            # Loop through segmentation frames
+            for out_seg, seg_frame_num in enumerate(frm_row):
+                if seg_frame_num >= 1:
+                    seg_frame_ind = seg_frame_num.item() - 1
+                    # Copy data to to output array
+                    if self.pixel_array.ndim == 2:
+                        # Special case with a single segmentation frame
+                        pixel_array[out_frm, :, :, out_seg] = \
+                            self.pixel_array
+                    else:
+                        pixel_array[out_frm, :, :, out_seg] = \
+                            self.pixel_array[seg_frame_ind, :, :]
+
+        if combine_segments:
+            # Check whether segmentation is binary, or fractional with only
+            # binary values
+            if self.segmentation_type != SegmentationTypeValues.BINARY:
+                is_binary = np.isin(
+                    np.unique(pixel_array),
+                    np.array([0.0, 1.0]),
+                    assume_unique=True
+                )
+                if not is_binary.all():
+                    raise ValueError(
+                        'Cannot combine segments if the segmentation is not'
+                        'binary'
+                    )
+
+            # Check for overlap by summing over the segments dimension
+            if np.any(pixel_array.sum(axis=-1) > 1):
+                raise RuntimeError(
+                    'Segments cannot be combined because they overlap'
+                )
+
+            # Scale the array by the segment number using broadcasting
+            if relabel:
+                pixel_value_map = np.arange(1, len(segment_numbers) + 1)
+            else:
+                pixel_value_map = segment_numbers
+            scaled_array = pixel_array * pixel_value_map.reshape(1, 1, 1, -1)
+
+            # Combine segments by taking maximum down final dimension
+            max_array = scaled_array.max(axis=-1)
+            pixel_array = max_array
+
+        return pixel_array
+
+    def get_source_sop_instances(self) -> List[Tuple[hd_UID, hd_UID, hd_UID]]:
```

Comment:
Since the method returns not only the SOP Instance UIDs of source images, I find the name misleading.

```suggestion
    def get_source_instance_uids(self) -> List[Tuple[hd_UID, hd_UID, hd_UID]]:
```